### PR TITLE
31 add authorizations and make sure logic is sound

### LINKED ISSUE #31: Add authorizations and make sure logic is sound

### PR DESCRIPTION
**Pull request by amandaag39**

### CODE REVIEW

File: app/controllers/analysis_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# app/controllers/open_ai_chats_controller.rb
+
+class AnalysisController < ApplicationController
+  before_action :authenticate_user!
```
Comment (thierrychau) [inline app/controllers/analysis_controller.rb]: is the intent to require users to authenticate only for accessing analysis?

File: app/controllers/analysis_controller.rb
Comment (thierrychau) [inline app/controllers/analysis_controller.rb:1]: nitpick: this should probably be in a separate PR.

File: app/controllers/readings_controller.rb [hunk 2]
```diff
@@ -70,17 +67,19 @@ def destroy
   end
 
   private
-    # Use callbacks to share common setup or constraints between actions.
+    def authorize_reading
+      authorize @reading
+    end
```
Comment (thierrychau) [inline app/controllers/readings_controller.rb]: personal preference: I think it's cleaner to call directly `authorize @reading` on line 4, as opposed to defining a new method.

File: app/controllers/readings_controller.rb [hunk 3]
```diff
@@ -1,9 +1,11 @@
 class ReadingsController < ApplicationController
+  before_action :authenticate_user!
   before_action :set_reading, only: %i[ show edit update destroy ]
+  before_action :authorize_reading, only: %i[show edit update destroy]
 
   # GET /readings or /readings.json
   def index
-    @readings = current_user.readings.order(measured_at: :asc)
+    @readings = policy_scope(Reading).order(measured_at: :asc)
```
Comment (thierrychau) [inline app/controllers/readings_controller.rb:7]: Good!

File: app/controllers/readings_controller.rb [hunk 4]
```diff
@@ -14,30 +16,25 @@ def index
...
   # POST /readings or /readings.json
   def create
-    # Assuming `vital_id` is passed as part of the form submission
     vital = current_user.vitals.find(params[:reading][:vital_id])
-  
-    # Initialize the reading with the user_id set to the current_user's id
     @reading = Reading.new(reading_params.merge(user_id: current_user.id, vital_id: vital.id))
   
+    authorize @reading
+
```
Comment (thierrychau) [inline app/controllers/readings_controller.rb:33]: refactor since you already have `before_action :authorize_reading, only: %i[show edit update destroy]` on line 4

File: app/controllers/readings_controller.rb [hunk 5]
```diff
@@ -14,30 +16,25 @@ def index
...
   end
 
-  # GET /readings/1 or /readings/1.json
-  def show
-  end
-
   # GET /readings/new
   def new
     @reading = Reading.new
+    authorize @reading, :new?
```
Comment (thierrychau) [inline app/controllers/readings_controller.rb]: refactor since you already have `before_action :authorize_reading, only: %i[show edit update destroy]` on line 4

File: app/controllers/vitals_controller.rb [hunk 6]
```diff
@@ -1,28 +1,23 @@
...
-
-  # GET /vitals/1/edit
-  def edit
+    authorize @vital
   end
 
   # POST /vitals or /vitals.json 
   def create
     @vital = current_user.vitals.new(vital_params)
+    authorize @vital
```
Comment (thierrychau) [inline app/controllers/vitals_controller.rb]: refactor

File: app/controllers/vitals_controller.rb [hunk 7]
```diff
@@ -1,28 +1,23 @@
...
   end
 
   # GET /vitals/new
   def new
     @vital = Vital.new
-  end
-
-  # GET /vitals/1/edit
-  def edit
+    authorize @vital
```
Comment (thierrychau) [inline app/controllers/vitals_controller.rb]: refactor

File: app/controllers/vitals_controller.rb [hunk 8]
```diff
@@ -85,38 +77,17 @@ def chart
...
-  # end
 
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_vital
       @vital = Vital.find(params[:id])
     end
 
+    def authorize_vital
+      authorize @vital
```
Comment (thierrychau) [inline app/controllers/vitals_controller.rb]: refactor

File: app/models/reading.rb [hunk 9]
```diff
@@ -20,6 +20,17 @@ class Reading < ApplicationRecord
...
+  scope :measured_at_between, ->(start_date, end_date) { where(measured_at: start_date..end_date) }
+  scope :measured_at_before, ->(end_date) { where("measured_at < ?", end_date) }
+  scope :measured_at_after, ->(start_date) { where("measured_at > ?", start_date) }
+  scope :for_vital_and_category, ->(name, category) {
+    joins(:vital).where(vitals: { name: name, category: Vital.categories[category] })
+  }
+  #scope :for_vital, -> (name, category) { joins(:vital).where( vitals: { name: name, category: category }) }
+  scope :for_numerical_prompt, -> { select(:measured_at, :numeric_reading) }
+  scope :for_text_prompt, -> { select(:measured_at, :text_reading) }
+  scope :default_order, -> { order('readings.measured_at DESC') }
```
Comment (thierrychau) [inline app/models/reading.rb:32]: nitpick: these should probably be in a different PR :)

File: app/models/user.rb [hunk 10]
```diff
@@ -18,4 +18,47 @@ class User < ApplicationRecord
...
+    Please give the patient one actionable step they can take that's simple to improve their situation. Analyze the following user data.
+    Numerical readings of #{numerical_vital_name} and textual descriptions of #{text_vital_name}:
+
+    #{numerical_vital_name} readings:
+    #{numerical_readings_formatted}
+
+    #{text_vital_name} descriptions:
+    #{text_readings_formatted}
+    PROMPT
+  end
```
Comment (thierrychau) [inline app/models/user.rb:63]: nitpick: these should probably be in a different PR :)

File: app/models/vital.rb [hunk 11]
```diff
@@ -12,7 +12,7 @@
 class Vital < ApplicationRecord
   enum category: { numerical: 0, text: 1, image: 2 }
   belongs_to :user
-  has_many :readings, dependent: :destroy
+  has_many :readings, dependent: :restrict_with_error #:destroy
```
Comment (thierrychau) [inline app/models/vital.rb:15]: nitpick: these should probably be in a different PR :)

File: app/policies/application_policy.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      @user = user
+      @record = record
+    end
+  
+    def index?
+      false
+    end
+  
+    def show?
+      scope.where(id: record.id).exists?
```
Comment (thierrychau) [inline app/policies/application_policy.rb]: What's the intent here?

File: app/policies/reading_policy.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
+# app/policies/reading_policy.rb
+class ReadingPolicy < ApplicationPolicy
+    class Scope < Scope
+      def resolve
+        scope.where(user_id: user.id)
+      end
+    end
+  
+    def create?
+        user.present?
```
Comment (thierrychau) [inline app/policies/reading_policy.rb]: because we force authentication for readings, this will already be true.

File: app/policies/reading_policy.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    def new?
+        create?
+    end
+
+    def show?
+      record.user_id == user.id
```
Comment (thierrychau) [inline app/policies/reading_policy.rb:22]: ok

File: app/policies/reading_policy.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  
+    def update?
+      record.user_id == user.id
```
Comment (thierrychau) [inline app/policies/reading_policy.rb:26]: ok

File: app/policies/reading_policy.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  
+    def destroy?
+      record.user_id == user.id
```
Comment (thierrychau) [inline app/policies/reading_policy.rb:30]: ok

File: app/services/open_ai_chat_service.rb
Comment (thierrychau) [inline app/services/open_ai_chat_service.rb:1]: nitpick: this should probably be in a separate PR.

File: app/controllers/analysis_controller.rb (hunk 1 above)
Comment (amandaag39) [reply app/controllers/analysis_controller.rb]: I see what you mean. I refactored to add ``before_action :authenticate_user!`` to the application_controller. I just added ``skip_before_action :authenticate_user!`` to the welcome controller since all users should see this regardless of whether or not they're logged in. Thanks!

File: app/controllers/readings_controller.rb (hunk 2 above)
Comment (amandaag39) [reply app/controllers/readings_controller.rb]: Hmm I'm not sure I understand what you mean here. Hopefully I have some time to ask you tomorrow!

File: app/controllers/readings_controller.rb (hunk 4 above)
Comment (amandaag39) [reply app/controllers/readings_controller.rb:33]: So, as I understand it, if I move this to the ``before_action :authorize_reading`` action, @reading is not yet set when the action starts since it depends on reading_params and other dynamic data that's not available until partway through the action. So if I call it before setting the @reading value (using params) it'll give me a nil value error. Maybe that's an issue with how I'm structuring my logic? It seems only to work if I place the ``authorize @reading`` after passing the dynamic param values into the @reading variable.

File: app/controllers/readings_controller.rb (hunk 2 above)
Comment (thierrychau) [reply app/controllers/readings_controller.rb]: Hmm, so you are defining a new method here called `authorize_reading` that returns `authorize @reading`. This saves 1 characters but I think add complexity to reading the code since now when I read line 4, I need to check the method you defined here to see what it does.

File: app/controllers/readings_controller.rb (hunk 4 above)
Comment (thierrychau) [reply app/controllers/readings_controller.rb:33]: I see, I think you are right. There may be a way to change how @reading is assigned here so it works. I'll look at this tomorrow.

File: app/controllers/readings_controller.rb (hunk 2 above)
Comment (amandaag39) [reply app/controllers/readings_controller.rb]: Ah! ok I think I see what you mean. So it's better to just pass it in as ``authorize @reading`` directly into the methods because making it a method doesn't make it more readable?

Comment (thierrychau) [reply app/controllers/readings_controller.rb]: Yes, just my personal preference. I may be wrong

File: app/controllers/readings_controller.rb (hunk 4 above)
Comment (amandaag39) [reply app/controllers/readings_controller.rb:33]: Ok thanks! Yeah no pressure to reply to me tonight 😅

Comment (thierrychau) [reply app/controllers/readings_controller.rb:33]: Actually, changing line 4 to do `before_action { authorize(@reading || Reading) }` would fix this `nil` value error, making the code a lot more DRY.

Basically, we are not really looking for an authorization rule on the specific instance of `@reading` here, so we can call `authorize` on the Class itself.

Comment (amandaag39) [reply app/controllers/readings_controller.rb:33]: Thank you! I was able to refactor using that. It also automatically gets rid of the authorize_reading method you mentioned not being necessary earlier. 😃

File: app/policies/reading_policy.rb (hunk 13 above)
Comment (amandaag39) [reply app/policies/reading_policy.rb]: So, for this policy I think I was trying to override the create method in the parent class (application policy) which looks like this:

```ruby
    def create?
      false
    end
 ```
 
I know it's redundant given the authorization in the controller, but if I don't override it, it defaults to a false value, so I was trying to make it secure by verifying if a user was present in the reading policy. Is it better to just set the application policy to true? Not sure how best to gracefully and securely handle that.

Comment (thierrychau) [reply app/policies/reading_policy.rb]: No, you are right for these 2 points:
- policies in application policy should **probably** always be set to false. This is the default behavior since all other policies inherit from it, and having it set to false means if we ever forget a policy, the worst case is nobody can access it.
- we do want to override individually for each policy and action.

Some nuance here that I think isn't 100% clear for you. Calling `authorize` in the controller only direct the controller to first check the policy available. If we don't call it, policies don't get applied. It does not in itself say who can do what, so it is not actually redundant. Does that make sense?

In this case, my comment was just to say if anyone logged in is allowed to perform this action, then you should probably just set it to `true` (which, in essence, is what you did by setting it to `user.present?` - but it's an added layer of abstraction)

Comment (amandaag39) [reply app/policies/reading_policy.rb]: Ok so that's why it defaults to false if I don't specify a policy in the reading policy class because 'authorize' in the controller looks for policy to match to determine whether or not something is authorized. It doesn't determine it on its own. I think I got it. Thank you!

File: app/policies/application_policy.rb (hunk 12 above)
Comment (amandaag39) [reply app/policies/application_policy.rb]: I think I was just trying to make sure the user can only see their own records, but yeah i see this is weird because I only did it for the show? and not for any other methods.  Also, as you mentioned above, it's probably better to just set this to false and let the reading or vital policies take care of it. I'll do that.
